proposed/activity-improvements.md

```diff
@@ -0,0 +1,465 @@
+    GetRequestedDataUsingParentId = (ref ActivityCreationOptions<string> options) => ...,
+    GetRequestedDataUsingContext = (ref ActivityCreationOptions<ActivityContext> options) => ... 
+};
+ActivitySource.AddActivityListener(listener);
+
+...
+    
+listener.Dispose();  
+```
+
+It is up to the listener to do extract information from the Activity object and respond to the callbacks however they like, probably by serializing a message to a remote service that will aggregate the telemetry.
+
+[TODO]: We could flesh out what we expect it look like end-to-end to use this with OpenTelemetry and perhaps some UI? It would help newcomers understand but I don't think it is a blocking issue for the current set of collaborators.
+
+Requirements:
+=============
+
+### Goals
+
+#### 1. All Activities can be collected
+
+The current design of DiagnosticListener and the conventions used by existing instrumented libraries expect the collection agent to have significant knowledge about each instrumented library. We need to make it possible for a general purpose collection agent to collect distributed trace telemetry despite having none of this advance knowledge. This also implies that any developer can add new Activity instrumentation to their app or library and the agent is capable or recording it automatically.
+
+#### 2. Activity collection can be filtered with a predicate on name
+
+We are assuming that most Activity information is desirable by default, but the monitoring agent should be able to filter out noisy or problematic instrumentation. The agent can efficiently decide if it wants to listen to Activities which created from specific source/Component (ActivitySource).
+
+#### 3. Activity needs API representations of OpenTelemetry Span concepts
+
+What .Net calls Activity, OpenTelemetry calls Span. The OT span can express a broader set of information than Activity currently can. We want to create parity so that developers who want to emit certain OpenTelemetry data points can still use Activity APIs to accomplish that.
+
+#### 4. API usability
+
+Some of the existing patterns of API usage are awkward and don't appeal to many developers. We'd like to simplify usage wherever possible so that the APIs feel more approachable to a broader set of .Net developers.
+
+#### 5. Back compat
+
+Any scenarios that currently work using Activity and/or DiagnosticListener must continue to work with no code changes. Also the performance for these scenarios can't meaningfully regress. Our new Activity listening mechanism should be faster than conventional listening.
+
+### Non-goals
+
+1. Creating interfaces so that custom implementations of Activity can be created (OpenTelemetry does this with Span)
+2. Supporting isolated monitoring domains that allows one set of instrumentation producers and consumers to be segregated from others within the same process (OpenTelemetry can do this via multiple instances of the top-level Tracer object)
```

Comment:
Would be great to clarify more on this point 2.

OT allows names Tracers. .NET allows named ActivitySource

In OT, all Tracer instances created from the same TracerProvider will have same listeners/exporter pipeline. And OT allows multiple TracerProvider to be present, each with separate pipeline.

In .NET, we'll only have single pipeline of listeners, enabled statically using `ActivitySource.AddActivityListener`. i.e activities from Every ActivitySource will be sent to the exact same set of listener.

Is the above correct? if yes, what is the reasoning behind not allowing the same capability as OT allows?

Reply:
The text needs to be revised a bit, partly due to my misunderstanding some of the capabilities of OT Tracer, partly a lack of precise language, and partly because the design+goals did evolve after it was written.

I think what would be accurate to say we only wanted to have a single Activity.Current and a single parent/child activity chain in each async ExecutionContext. Imagine some code like this:

````
ActivitySource sourceA = … 
ActivitySource sourceB = …
using (var a = sourceA.Start(...))
{
    using (var b = sourceB.Start(…)) {}
} 
````

We support any number of listeners and each of them can independently choose to listen to A, B, both, or neither. However if there are listeners for A and B then A will always be the parent of B. In a more complex system that we are not aiming to support you could imagine that sourceA and sourceB are completely isolated from each other so that instead of A being the parent of B, there are two chains of activities being tracked in parallel with A as the root of one and B as the root of a 2nd. Rather than having Activity.Current, now you would have domain A's current activity and domain B's current Activity.

> i.e activities from Every ActivitySource will be sent to the exact same set of listener.

Each listener can choose to subscribe to an arbitrary subset of the ActivitySources using the ShouldListenTo(ActivitySource source) callback. Activities from an ActivitySource will only be sent to the listeners which subscribed to it.

Does that help clarify it? If it did, do you think we've supported enough functionality or you are concerned we aren't going farther?

Reply:
Thanks for the explanation for the 1st part - about Activit.Current. This makes total sense to me.

Reply:
The 2nd part - about activities from every ActivitySource is sent to all listeners - I was bit confused on this part, in comparison with OT spec:

In OT, there can be multiple TracerProviders, each with its own processing pipelines (Listeners)
So its possible for one to construct a SecretTracerProvider, which has its own SecretPipeline, and a DefaultTracerProvider, with DefaultPipeline.
If a library handling privacy data uses SecretTracerProvider, traces emitted from the library will **only** be sent to the SecretPipeline.

In case of Activity, the above scenario is not achievable as every ActivityListener gets all the activity, but they can chose to ignore it. But nothing prevents the listener from listening on all of them, including secrets.

I was asking the above from a privacy perspective. @reyang shared his thoughts about this as well - basically everything emitted in-proc is accessible to all of execution context, and within the process address space, there is no security.

Not a blocker, just listing down observations :)

Reply:
Agreed, the current design won't support being able to create a secret pipeline. It assumes all listeners cooperatively enforce privacy boundaries or that no privacy boundary needs to be enforced.